### PR TITLE
Add an endpoint that allows admins to assign the `memberTo` field from a sheet from THS

### PR DESCRIPTION
**Pull request by foodelevator**
It would be nice to also have a simple frontend for this, but in my testing I have used:

```sh
curl http://localhost:8080/membership-sheet -X POST \
  -H "X-Api-Key: $HODIS_API_KEY" \
  --data-binary @$HOME/Downloads/Medlemslista_2.xlsx
```

### CODE REVIEW

File: .github/workflows/main.yaml [hunk 1]
```diff
@@ -16,32 +16,32 @@ jobs:
     - name: Set up Go 1.x
       uses: actions/setup-go@v2
       with:
-        go-version: ^1.15
+        go-version: ^1.20
```
Comment (Herkarl) [inline .github/workflows/main.yaml:19]: ballsy

File: handlers/handlers.go [hunk 2]
```diff
@@ -90,6 +94,132 @@ func Update(db *gorm.DB) gin.HandlerFunc {
...
+					errorRows = append(errorRows, email)
+				}
+				continue
+			}
+
+			memberTo, err := time.Parse(time.DateOnly, date)
+
+			// Note that although very common, everyone's kth email is not their kth id followed by
+			// @kth.se. KTH:s ldap can however not be searched by emails, so I don't know of a good
+			// solution.
```
Comment (Herkarl) [inline handlers/handlers.go:201]: When reading this I get a bit worried about how THS handles support-members once their mail get deleted. Will they have an empty row in the email slot, or will they have their old kth-email which does not exist anymore (or possibly has been assigned to someone new)?

I'm unsure if KTH has actually started deleting emails yet, so I don't know if there are any examples we can look at.

Very much a special case nonthenless, so perhaps not that important..

File: handlers/handlers.go [hunk 3]
```diff
@@ -156,8 +293,13 @@ func Authenticate(loginURL, apiKey string) gin.HandlerFunc {
...
 }
 
-func HasPlsPermission(tokenType, tokenValue, permission string) bool {
-	url := fmt.Sprintf("https://pls.datasektionen.se/api/%s/%s/hodis/%s", tokenType, tokenValue, permission)
+func HasPlsPermission(plsURL, tokenType, tokenValue, permission string) bool {
+	if plsURL == "true" {
+		return true
+	} else if plsURL == "false" {
+		return false
+	}
```
Comment (Herkarl) [inline handlers/handlers.go:307]: nice :stuck_out_tongue:

File: handlers/handlers.go [hunk 4]
```diff
@@ -90,6 +94,132 @@ func Update(db *gorm.DB) gin.HandlerFunc {
...
+		columns, err := rows.Columns()
+		if err != nil {
+			c.JSON(400, gin.H{"error": err.Error()})
+			c.Abort()
+			return
+		}
+		var dateCol, emailCol, chapterCol int = -1, -1, -1
+		for i, title := range columns {
+			title = strings.TrimSpace(title)
+			if title == "Giltig till" {
```
Comment (Herkarl) [inline handlers/handlers.go]: Could be nice to have these as constants instead of inline.

**Review by Herkarl — APPROVED** [review]
lgtm overall, had some thoughts while looking through, but ready to merge imo.

As an alternative to matching column titles with hard-coded strings, I think it could be nice to have the request provide the titles, or maybe even the column indices (probably more robust).

Then you could have some nice interface for inputting those in the frontend. A feature creep idea could be to even have a preview for the matrix and then have the user select the columns in that view.

But definitely not something that is needed for now :stuck_out_tongue:, just an idea for future expansion.

File: handlers/handlers.go (hunk 2 above)
Comment (foodelevator) [reply handlers/handlers.go:201]: Definitely good to think about but hopefully we don't need to!

**Pull request merged**